File: Tests/SentryTests/SentryCrash/SentryCrashBinaryImageCacheTests.m
#import "SentryCrashBinaryImageCache+Test.h"
#import "SentryCrashBinaryImageCache.h"
#import "SentryCrashDynamicLinker+Test.h"
#import "SentrySwift.h"
#import <XCTest/XCTest.h>

#include <mach-o/dyld.h>
#include <mach-o/dyld_images.h>

// Test-only functions are declared in `SentryCrashBinaryImageCache+Test.h`

static void (*addBinaryImage)(const struct mach_header *mh, intptr_t vmaddr_slide);
static void (*removeBinaryImage)(const struct mach_header *mh, intptr_t vmaddr_slide);
static NSMutableArray *mach_headers_test_cache;
static NSMutableArray *mach_headers_expect_array;
static const uint32_t maxDyldImages = 4096;

static void
sentry_register_func_for_add_image(
    void (*func)(const struct mach_header *mh, intptr_t vmaddr_slide))
{
    addBinaryImage = func;

    if (mach_headers_expect_array) {
        // Skipping first item which is dyld and already included when starting the cache
        for (NSUInteger i = 1; i < mach_headers_expect_array.count; i++) {
            NSValue *header = mach_headers_expect_array[i];
            func(header.pointerValue, 0);
        }
    }
}

static void
sentry_register_func_for_remove_image(
    void (*func)(const struct mach_header *mh, intptr_t vmaddr_slide))
{
    removeBinaryImage = func;
}

static void
cacheMachHeaders(const struct mach_header *mh, __unused intptr_t vmaddr_slide)
{
    [mach_headers_test_cache addObject:[NSValue valueWithPointer:mh]];
}

static void
countNumberOfImagesInCache(__unused SentryCrashBinaryImage *image, void *context)
{
    int *counter = context;
    (*counter)++;
}

static void
addBinaryImageToArray(SentryCrashBinaryImage *image, void *context)
{
    NSMutableArray *array = (__bridge NSMutableArray *)context;
    [array addObject:[NSValue valueWithPointer:image]];
}

static void
addBinaryImageNameToArray(SentryCrashBinaryImage *image, void *context)
{
    NSMutableArray *array = (__bridge NSMutableArray *)context;
    if (image->name) {
        [array addObject:[NSString stringWithUTF8String:image->name]];
    } else {
        [array addObject:@"<null>"];
    }
}

static NSMutableArray<NSString *> *added_image_names;

static void
captureAddedImageName(const SentryCrashBinaryImage *image)
{
    @synchronized(added_image_names) {
        if (image->name) {
            [added_image_names addObject:[NSString stringWithUTF8String:image->name]];
        } else {
            [added_image_names addObject:@"<null>"];
        }
    }
}

static NSArray<NSString *> *
copyAddedImageNames(void)
{
    @synchronized(added_image_names) {
        return [added_image_names copy] ?: @[];
    }
}

dispatch_semaphore_t delaySemaphore = NULL;
dispatch_semaphore_t delayCalled = NULL;
static void
delayAddBinaryImage(void)
{
    if (delayCalled) {
        dispatch_semaphore_signal(delayCalled);
    }
    if (delaySemaphore) {
        dispatch_semaphore_wait(delaySemaphore, DISPATCH_TIME_FOREVER);
    }
}

@interface SentryCrashBinaryImageCacheTests : XCTestCase

@end

@implementation SentryCrashBinaryImageCacheTests

+ (void)setUp
{
    // Create a test cache of actual binary images to be used during tests.
    mach_headers_test_cache = [NSMutableArray array];

    // Manually include dyld
    sentrycrashdl_initialize();
    [mach_headers_test_cache addObject:[NSValue valueWithPointer:sentryDyldHeader]];
    _dyld_register_func_for_add_image(&cacheMachHeaders);
}

- (void)setUp
{
    sentrycrashbic_useFreshTestCacheState();
    sentrycrashbic_setRegisterFuncForAddImage(&sentry_register_func_for_add_image);
    sentrycrashbic_setRegisterFuncForRemoveImage(&sentry_register_func_for_remove_image);

    added_image_names = [NSMutableArray array];
    delaySemaphore = NULL;
    delayCalled = NULL;

    // Copying the first 5 images from the temporary list.
    // 5 is a magic number.
    mach_headers_expect_array =
        [mach_headers_test_cache subarrayWithRange:NSMakeRange(0, 5)].mutableCopy;
}

- (void)tearDown
{
    added_image_names = nil;
    delaySemaphore = NULL;
    delayCalled = NULL;

    sentrycrashdl_clearDyld();
    sentrycrashbic_useDefaultCacheState();
    [SentryDependencyContainer reset];
}

- (void)testStartCache
{
    [SentryDependencyContainer.sharedInstance.crashWrapper startBinaryImageCache];
    [self assertBinaryImageCacheLength:5];
}

- (void)testStartCacheTwice
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];
}

- (void)testStopCache
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];
    sentrycrashbic_stopCache();
    [self assertBinaryImageCacheLength:5];
}

- (void)testStopCacheTwice
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];
    sentrycrashbic_stopCache();
    [self assertBinaryImageCacheLength:5];
    sentrycrashbic_stopCache();
    [self assertBinaryImageCacheLength:5];
}

- (void)testRegisterAddedCallbackBeforeStartingCache
{
    // Deterministically simulates the async production path where the callback is installed before
    // `addDyldImage()` runs.
    sentrycrashbic_registerAddedCallback(&captureAddedImageName);

    sentrycrashbic_startCache();

    NSArray<NSString *> *names = copyAddedImageNames();
    XCTAssertEqual(5, names.count);
    XCTAssertTrue([names containsObject:@"dyld"]);
}

- (void)testUseFreshTestCacheState_whenSwitchingStates_shouldResetAndIsolateTheActiveCache
{
    sentrycrashbic_useDefaultCacheState();
    sentrycrashbic_setRegisterFuncForAddImage(&sentry_register_func_for_add_image);
    sentrycrashbic_setRegisterFuncForRemoveImage(&sentry_register_func_for_remove_image);

    sentrycrashbic_startCache();
    addBinaryImage([mach_headers_test_cache[5] pointerValue], 0);
    [self assertBinaryImageCacheLength:6];

    sentrycrashbic_useFreshTestCacheState();
    sentrycrashbic_setRegisterFuncForAddImage(&sentry_register_func_for_add_image);
    sentrycrashbic_setRegisterFuncForRemoveImage(&sentry_register_func_for_remove_image);

    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    addBinaryImage([mach_headers_test_cache[6] pointerValue], 0);
    [self assertBinaryImageCacheLength:6];

    sentrycrashbic_useDefaultCacheState();
    sentrycrashbic_setRegisterFuncForAddImage(&sentry_register_func_for_add_image);
    sentrycrashbic_setRegisterFuncForRemoveImage(&sentry_register_func_for_remove_image);

    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];
}

- (void)testAddNewImage
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    addBinaryImage([mach_headers_test_cache[5] pointerValue], 0);
    mach_headers_expect_array =
        [mach_headers_test_cache subarrayWithRange:NSMakeRange(0, 6)].mutableCopy;
    [self assertBinaryImageCacheLength:6];
    [self assertCachedBinaryImages];

    addBinaryImage([mach_headers_test_cache[6] pointerValue], 0);
    mach_headers_expect_array =
        [mach_headers_test_cache subarrayWithRange:NSMakeRange(0, 7)].mutableCopy;
    [self assertBinaryImageCacheLength:7];
    [self assertCachedBinaryImages];
}

- (void)testAddInvalidHeader
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    addBinaryImage(0, 0);
    [self assertBinaryImageCacheLength:5];
}

- (void)testAddNewImageAfterStopping
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    sentrycrashbic_stopCache();
    addBinaryImage([mach_headers_test_cache[6] pointerValue], 0);
    [self assertBinaryImageCacheLength:6];
}

- (void)testRemoveImageFromTail
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    removeBinaryImage([mach_headers_expect_array[4] pointerValue], 0);
    [self assertBinaryImageCacheLength:4];
    [self assertCachedBinaryImages];

    removeBinaryImage([mach_headers_expect_array[3] pointerValue], 0);
    [self assertBinaryImageCacheLength:3];
    [self assertCachedBinaryImages];
}

- (void)testRemoveImageFromBeginning
{
    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    removeBinaryImage([mach_headers_expect_array[0] pointerValue], 0);
    [self assertBinaryImageCacheLength:4];
    [mach_headers_expect_array removeObjectAtIndex:0];
    [self assertCachedBinaryImages];

    removeBinaryImage([mach_headers_expect_array[0] pointerValue], 0);
    [self assertBinaryImageCacheLength:3];
    [mach_headers_expect_array removeObjectAtIndex:0];
    [self assertCachedBinaryImages];
}

- (void)testRemoveImageAddAgain
{
    // Use index 1 since we can't dynamically insert dyld image (`dladdr` returns null)
    int indexToRemove = 1;

    sentrycrashbic_startCache();
    [self assertBinaryImageCacheLength:5];

    removeBinaryImage([mach_headers_expect_array[indexToRemove] pointerValue], 0);
    [self assertBinaryImageCacheLength:4];

    NSValue *removeItem = mach_headers_expect_array[indexToRemove];
    [mach_headers_expect_array removeObjectAtIndex:indexToRemove];
    [self assertCachedBinaryImages];

    addBinaryImage(removeItem.pointerValue, 0);
    [self assertBinaryImageCacheLength:5];
    [mach_headers_expect_array insertObject:removeItem atIndex:4];
    [self assertCachedBinaryImages];
}

- (void)testAddBinaryImageInParallel
{
    sentrycrashbic_startCache();
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);

    // Guard against underflow when mach_headers_test_cache.count < 5
    // because otherwise the expectedFulfillmentCount for the test expectation will be negative.
    NSInteger taskCount = mach_headers_test_cache.count - 5;
    if (taskCount <= 0) {
        XCTFail(@"Expected a positive task count, but got %ld", taskCount);
        return;
    }

    XCTestExpectation *expectation =
        [self expectationWithDescription:@"Add binary images in parallel"];
    expectation.expectedFulfillmentCount = taskCount;

    for (NSUInteger i = 5; i < mach_headers_test_cache.count; i++) {
        dispatch_async(queue, ^{
            addBinaryImage([mach_headers_test_cache[i] pointerValue], 0);
            [expectation fulfill];
        });
    }

    [self waitForExpectations:@[ expectation ] timeout:5.0];

    [self assertBinaryImageCacheLength:(int)mach_headers_test_cache.count];
}

- (void)testCloseCacheWhileAdding
{
    sentrycrashbic_startCache();
    sentrycrashbic_setBeforeAddImageCallback(&delayAddBinaryImage);
    delaySemaphore = dispatch_semaphore_create(0);
    delayCalled = dispatch_semaphore_create(0);
    dispatch_semaphore_t addFinished = dispatch_semaphore_create(0);

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        addBinaryImage([mach_headers_test_cache[6] pointerValue], 0);
        dispatch_semaphore_signal(addFinished);
    });

    intptr_t result
        = dispatch_semaphore_wait(delayCalled, dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC));
    sentrycrashbic_stopCache();
    dispatch_semaphore_signal(delaySemaphore);
    intptr_t addResult
        = dispatch_semaphore_wait(addFinished, dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC));
    [self assertBinaryImageCacheLength:6];
    XCTAssertEqual(result, 0);
    XCTAssertEqual(addResult, 0);
}

// Adding a SentryBinaryImageCache test inside
// SentryCrashBinaryImageCache to test integration between both
// because is easier to control SentryCrashBinaryImageCache in an objc test
- (void)testSentryBinaryImageCacheIntegration
{
    sentrycrashbic_startCache();

    SentryBinaryImageCache *imageCache = SentryDependencyContainer.sharedInstance.binaryImageCache;
    [imageCache start:false];
    // by calling start, SentryBinaryImageCache will register a callback with
    // `SentryCrashBinaryImageCache` that should be called for every image already cached.
    NSMutableArray<NSString *> *paths = [NSMutableArray new];
    [imageCache.cache enumerateObjectsUsingBlock:^(SentryBinaryImageInfo *_Nonnull obj,
        NSUInteger __unused idx, BOOL *_Nonnull __unused stop) { [paths addObject:obj.name]; }];
    XCTAssertEqual(
        5, imageCache.cache.count, @"Cache should start with 5 images but contained %@", paths);

    addBinaryImage([mach_headers_test_cache[5] pointerValue], 0);
    XCTAssertEqual(6, imageCache.cache.count);

    removeBinaryImage([mach_headers_expect_array[1] pointerValue], 0);
    removeBinaryImage([mach_headers_expect_array[2] pointerValue], 0);
    XCTAssertEqual(4, imageCache.cache.count);
    [imageCache stop];

    addBinaryImage([mach_headers_test_cache[6] pointerValue], 0);
    XCTAssertNil(imageCache.cache);
}

- (void)testImagesAddedWhileSwiftCacheStopped_AreReplayedOnRestart
{
    sentrycrashbic_startCache();

    SentryBinaryImageCache *imageCache = SentryDependencyContainer.sharedInstance.binaryImageCache;
    [imageCache start:false];
    XCTAssertEqual(5, imageCache.cache.count);

    [imageCache stop];
    addBinaryImage([mach_headers_test_cache[5] pointerValue], 0);
    XCTAssertNil(imageCache.cache);

    [imageCache start:false];
    XCTAssertEqual(6, imageCache.cache.count);
}

- (void)assertBinaryImageCacheLength:(int)expected
{
    int counter = 0;
    sentrycrashbic_iterateOverImages(countNumberOfImagesInCache, &counter);
    NSMutableArray<NSString *> *names = [NSMutableArray new];
    sentrycrashbic_iterateOverImages(addBinaryImageNameToArray, (__bridge void *)(names));
    XCTAssertEqual(
        counter, expected, @"Cache should have %d images but contained %@", expected, names);
}

- (void)assertCachedBinaryImages
{
    NSArray *cached = [self binaryImageCacheToArray];
    for (NSUInteger i = 0; i < cached.count; i++) {
        SentryCrashBinaryImage *binaryImage = [cached[i] pointerValue];
        struct mach_header *header = [mach_headers_expect_array[i] pointerValue];
        XCTAssertEqual(binaryImage->address, (uint64_t)header);
    }
}

- (NSArray *)binaryImageCacheToArray
{
    NSMutableArray *result = [NSMutableArray array];
    sentrycrashbic_iterateOverImages(addBinaryImageToArray, (__bridge void *)(result));
    return result;
}

- (const struct mach_header *)capacityTestHeader
{
    return [mach_headers_expect_array[1] pointerValue];
}

- (void)bootstrapDyldCallbacksAndResetCacheState
{
    sentrycrashbic_startCache();
    sentrycrashbic_useFreshTestCacheState();
    sentrycrashbic_setRegisterFuncForAddImage(&sentry_register_func_for_add_image);
    sentrycrashbic_setRegisterFuncForRemoveImage(&sentry_register_func_for_remove_image);
}

- (void)fillStartedCacheToImageLimitWithHeader:(const struct mach_header *)header
{
    for (uint32_t i = 5; i < maxDyldImages; i++) {
        addBinaryImage(header, 0);
    }
}

- (void)fillInactiveCacheToImageLimitWithHeader:(const struct mach_header *)header
{
    for (uint32_t i = 0; i < maxDyldImages; i++) {
        addBinaryImage(header, 0);
    }
}

- (void)testAddImage_whenCacheIsAtCapacity_shouldIgnoreOverflowImage
{
    const struct mach_header *header = [self capacityTestHeader];

    sentrycrashbic_startCache();
    [self fillStartedCacheToImageLimitWithHeader:header];
    [self assertBinaryImageCacheLength:maxDyldImages];

    sentrycrashbic_registerAddedCallback(&captureAddedImageName);
    XCTAssertEqual((NSUInteger)maxDyldImages, copyAddedImageNames().count);

    addBinaryImage(header, 0);

    XCTAssertEqual((NSUInteger)maxDyldImages, copyAddedImageNames().count);
    [self assertBinaryImageCacheLength:maxDyldImages];
}

- (void)testStartCache_whenCacheIsAtCapacity_shouldSkipAddingDyld
{
    const struct mach_header *header = [self capacityTestHeader];

    [self bootstrapDyldCallbacksAndResetCacheState];
    [self fillInactiveCacheToImageLimitWithHeader:header];

    NSMutableArray<NSString *> *namesBeforeStart = [NSMutableArray array];
    sentrycrashbic_iterateOverImages(
        addBinaryImageNameToArray, (__bridge void *)(namesBeforeStart));
    XCTAssertEqual((NSUInteger)maxDyldImages, namesBeforeStart.count);
    XCTAssertFalse([namesBeforeStart containsObject:@"dyld"]);

    sentrycrashbic_startCache();

    NSMutableArray<NSString *> *namesAfterStart = [NSMutableArray array];
    sentrycrashbic_iterateOverImages(addBinaryImageNameToArray, (__bridge void *)(namesAfterStart));
    XCTAssertEqual((NSUInteger)maxDyldImages, namesAfterStart.count);
    XCTAssertFalse([namesAfterStart containsObject:@"dyld"]);
}

- (void)testIterateOverImages_whenNextIndexExceedsMax_shouldIgnoreOverflowSlot
{
    const struct mach_header *header = [self capacityTestHeader];

    [self bootstrapDyldCallbacksAndResetCacheState];
    [self fillInactiveCacheToImageLimitWithHeader:header];
    addBinaryImage(header, 0);

    XCTAssertEqual((NSUInteger)maxDyldImages, [self binaryImageCacheToArray].count);
}

- (void)testRegisterAddedCallback_whenNextIndexExceedsMax_shouldReplayOnlyTrackedImages
{
    const struct mach_header *header = [self capacityTestHeader];

    [self bootstrapDyldCallbacksAndResetCacheState];
    [self fillInactiveCacheToImageLimitWithHeader:header];
    addBinaryImage(header, 0);

    sentrycrashbic_registerAddedCallback(&captureAddedImageName);

    NSArray<NSString *> *replayedNames = copyAddedImageNames();
    XCTAssertEqual((NSUInteger)maxDyldImages, replayedNames.count);
    XCTAssertFalse([replayedNames containsObject:@"dyld"]);
}

- (void)testRemoveImage_whenNextIndexExceedsMax_shouldRemoveTrackedImage
{
    const struct mach_header *header = [self capacityTestHeader];

    [self bootstrapDyldCallbacksAndResetCacheState];
    [self fillInactiveCacheToImageLimitWithHeader:header];
    addBinaryImage(header, 0);

    removeBinaryImage(header, 0);

    [self assertBinaryImageCacheLength:maxDyldImages - 1];
}

@end

File: Tests/SentryTests/SentryCrash/SentryCrashBinaryImageCacheTests.swift
import MachO
@_spi(Private) @testable import Sentry
@_spi(Private) @testable import SentryTestUtils
import XCTest

// Import C functions from the SentryCrashBinaryImageCache
// These are test-only functions exposed from the implementation
@_silgen_name("sentry_setRegisterFuncForAddImage")
func sentry_setRegisterFuncForAddImage(_ addFunction: UnsafeMutableRawPointer?)

@_silgen_name("sentry_setRegisterFuncForRemoveImage") 
func sentry_setRegisterFuncForRemoveImage(_ removeFunction: UnsafeMutableRawPointer?)

@_silgen_name("sentry_resetFuncForAddRemoveImage")
func sentry_resetFuncForAddRemoveImage()

@_silgen_name("sentry_setFuncForBeforeAdd")
func sentry_setFuncForBeforeAdd(_ callback: (@convention(c) () -> Void)?)

@_silgen_name("sentrycrashbic_iterateOverImages")
func sentrycrashbic_iterateOverImages(_ callback: (@convention(c) (UnsafeMutablePointer<SentryCrashBinaryImage>, UnsafeMutableRawPointer?) -> Void)?, _ context: UnsafeMutableRawPointer?)

@_silgen_name("sentrycrashdl_clearDyld")
func sentrycrashdl_clearDyld()

// Global variables to mirror the Objective-C implementation
private var addBinaryImageFunc: (@convention(c) (UnsafePointer<mach_header>?, Int) -> Void)?
private var removeBinaryImageFunc: (@convention(c) (UnsafePointer<mach_header>?, Int) -> Void)?
private var machHeadersTestCache: [UnsafeRawPointer] = []
private var machHeadersExpectArray: [UnsafeRawPointer] = []

// Semaphores for testing parallel operations
private var delaySemaphore: DispatchSemaphore?
private var delayCalled: DispatchSemaphore?

// C callback functions
private let sentryRegisterFuncForAddImage: @convention(c) ((@convention(c) (UnsafePointer<mach_header>?, Int) -> Void)?) -> Void =
{ function in
    addBinaryImageFunc = function
    
    if !machHeadersExpectArray.isEmpty {
        // Skip first item which is dyld and already included when starting the cache
        for i in 1..<machHeadersExpectArray.count {
            let header = machHeadersExpectArray[i].assumingMemoryBound(to: mach_header.self)
            function?(header, 0)
        }
    }
}

private let sentryRegisterFuncForRemoveImage: @convention(c) ((@convention(c) (UnsafePointer<mach_header>?, Int) -> Void)?) -> Void = { function in
    removeBinaryImageFunc = function
}

private let cacheMachHeaders: @convention(c) (UnsafePointer<mach_header>?, Int) -> Void = { mh, _ in
    guard let mh = mh else { return }
    machHeadersTestCache.append(UnsafeRawPointer(mh))
}

private let countNumberOfImagesInCache: @convention(c) (UnsafeMutablePointer<SentryCrashBinaryImage>, UnsafeMutableRawPointer?) -> Void = { _, context in
    guard let context = context else { return }
    let counter = context.assumingMemoryBound(to: Int.self)
    counter.pointee += 1
}

private let addBinaryImageToArray: @convention(c) (UnsafeMutablePointer<SentryCrashBinaryImage>, UnsafeMutableRawPointer?) -> Void = { image, context in
    guard let context = context else { return }
    let array = Unmanaged<NSMutableArray>.fromOpaque(context).takeUnretainedValue()
    array.add(NSValue(pointer: image))
}

private let delayAddBinaryImage: @convention(c) () -> Void = {
    if let delayCalled = delayCalled {
        delayCalled.signal()
    }
    if let delaySemaphore = delaySemaphore {
        delaySemaphore.wait()
    }
}

class SentryCrashBinaryImageCacheTests: XCTestCase {
    
    override class func setUp() {
        super.setUp()
        // Create a test cache of actual binary images to be used during tests
        machHeadersTestCache = []
        
        // Manually include dyld
        sentrycrashdl_initialize()
        if let dyldHeader = sentryDyldHeader {
            machHeadersTestCache.append(UnsafeRawPointer(dyldHeader))
        }
        _dyld_register_func_for_add_image(cacheMachHeaders)
    }
    
    override func setUp() {
        super.setUp()
        sentry_setRegisterFuncForAddImage(unsafeBitCast(sentryRegisterFuncForAddImage, to: UnsafeMutableRawPointer.self))
        sentry_setRegisterFuncForRemoveImage(unsafeBitCast(sentryRegisterFuncForRemoveImage, to: UnsafeMutableRawPointer.self))
        
        // Copy the first 5 images from the temporary list
        // 5 is a magic number from the original implementation
        let endIndex = min(5, machHeadersTestCache.count)
        machHeadersExpectArray = Array(machHeadersTestCache[0..<endIndex])
    }
    
    override func tearDown() {
        sentrycrashdl_clearDyld()
        sentry_resetFuncForAddRemoveImage()
        sentrycrashbic_stopCache()
        sentry_setFuncForBeforeAdd(nil)
        SentryDependencyContainer.reset()
        super.tearDown()
    }
    
    func testStartCache() {
        SentryCrashWrapper.sharedInstance().startBinaryImageCache()
        assertBinaryImageCacheLength(5)
    }
    
    func testStartCacheTwice() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
    }
    
    func testStopCache() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        sentrycrashbic_stopCache()
        assertBinaryImageCacheLength(0)
    }
    
    func testStopCacheTwice() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        sentrycrashbic_stopCache()
        assertBinaryImageCacheLength(0)
        sentrycrashbic_stopCache()
        assertBinaryImageCacheLength(0)
    }
    
    func testAddNewImage() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        if machHeadersTestCache.count > 5 {
            let header = machHeadersTestCache[5].assumingMemoryBound(to: mach_header.self)
            addBinaryImageFunc?(header, 0)
            machHeadersExpectArray = Array(machHeadersTestCache[0..<6])
            assertBinaryImageCacheLength(6)
            assertCachedBinaryImages()
        }
        
        if machHeadersTestCache.count > 6 {
            let header = machHeadersTestCache[6].assumingMemoryBound(to: mach_header.self)
            addBinaryImageFunc?(header, 0)
            machHeadersExpectArray = Array(machHeadersTestCache[0..<7])
            assertBinaryImageCacheLength(7)
            assertCachedBinaryImages()
        }
    }
    
    func testAddInvalidHeader() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        addBinaryImageFunc?(nil, 0)
        assertBinaryImageCacheLength(5)
    }
    
    func testAddNewImageAfterStopping() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        sentrycrashbic_stopCache()
        if machHeadersTestCache.count > 6 {
            let header = machHeadersTestCache[6].assumingMemoryBound(to: mach_header.self)
            addBinaryImageFunc?(header, 0)
        }
        assertBinaryImageCacheLength(0)
    }
    
    func testRemoveImageFromTail() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        let header4 = machHeadersExpectArray[4].assumingMemoryBound(to: mach_header.self)
        removeBinaryImageFunc?(header4, 0)
        assertBinaryImageCacheLength(4)
        assertCachedBinaryImages()
        
        let header3 = machHeadersExpectArray[3].assumingMemoryBound(to: mach_header.self)
        removeBinaryImageFunc?(header3, 0)
        assertBinaryImageCacheLength(3)
        assertCachedBinaryImages()
    }
    
    func testRemoveImageFromBeginning() {
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        let header0 = machHeadersExpectArray[0].assumingMemoryBound(to: mach_header.self)
        removeBinaryImageFunc?(header0, 0)
        assertBinaryImageCacheLength(4)
        machHeadersExpectArray.removeFirst()
        assertCachedBinaryImages()
        
        let newHeader0 = machHeadersExpectArray[0].assumingMemoryBound(to: mach_header.self)
        removeBinaryImageFunc?(newHeader0, 0)
        assertBinaryImageCacheLength(3)
        machHeadersExpectArray.removeFirst()
        assertCachedBinaryImages()
    }
    
    func testRemoveImageAddAgain() {
        // Use index 1 since we can't dynamically insert dyld image (`dladdr` returns null)
        let indexToRemove = 1
        
        sentrycrashbic_startCache()
        assertBinaryImageCacheLength(5)
        
        let headerToRemove = machHeadersExpectArray[indexToRemove].assumingMemoryBound(to: mach_header.self)
        removeBinaryImageFunc?(headerToRemove, 0)
        assertBinaryImageCacheLength(4)
        
        let removeItem = machHeadersExpectArray[indexToRemove]
        machHeadersExpectArray.remove(at: indexToRemove)
        assertCachedBinaryImages()
        
        addBinaryImageFunc?(removeItem.assumingMemoryBound(to: mach_header.self), 0)
        assertBinaryImageCacheLength(5)
        machHeadersExpectArray.insert(removeItem, at: 4)
        assertCachedBinaryImages()
    }
    
    func testAddBinaryImageInParallel() {
        sentrycrashbic_startCache()
        let queue = DispatchQueue.global(qos: .default)
        
        // Guard against underflow when machHeadersTestCache.count < 5
        let taskCount = machHeadersTestCache.count - 5
        guard taskCount > 0 else {
            XCTFail("Expected a positive task count, but got \(taskCount)")
            return
        }
        
        let expectation = self.expectation(description: "Add binary images in parallel")
        expectation.expectedFulfillmentCount = taskCount
        
        for i in 5..<machHeadersTestCache.count {
            queue.async {
                let header = machHeadersTestCache[i].assumingMemoryBound(to: mach_header.self)
                addBinaryImageFunc?(header, 0)
                expectation.fulfill()
            }
        }
        
        waitForExpectations(timeout: 5.0)
        assertBinaryImageCacheLength(machHeadersTestCache.count)
    }
    
    func testCloseCacheWhileAdding() {
        sentrycrashbic_startCache()
        sentry_setFuncForBeforeAdd(delayAddBinaryImage)
        delaySemaphore = DispatchSemaphore(value: 0)
        delayCalled = DispatchSemaphore(value: 0)
        
        DispatchQueue.global(qos: .default).async {
            if machHeadersTestCache.count > 6 {
                let header = machHeadersTestCache[6].assumingMemoryBound(to: mach_header.self)
                addBinaryImageFunc?(header, 0)
            }
        }
        
        let result = delayCalled?.wait(timeout: .now() + 5)
        sentrycrashbic_stopCache()
        delaySemaphore?.signal()
        assertBinaryImageCacheLength(0)
        XCTAssertEqual(result, .success)
    }
    
    // Adding a SentryBinaryImageCache test inside SentryCrashBinaryImageCache to test integration 
    // The test is in Swift because of some classes are not available to Objective-C
    func testSentryBinaryImageCacheIntegration() {
        sentrycrashbic_startCache()
        
        let imageCache = SentryDependencyContainer.sharedInstance().binaryImageCache
        let dispatchQueueWrapper = TestSentryDispatchQueueWrapper()
        imageCache.start(false, dispatchQueueWrapper: dispatchQueueWrapper)
        // By calling start, SentryBinaryImageCache will register a callback with
        // `SentryCrashBinaryImageCache` that should be called for every image already cached.
        XCTAssertEqual(5, imageCache.cache?.count ?? 0)
        
        if machHeadersTestCache.count > 5 {
            let header = machHeadersTestCache[5].assumingMemoryBound(to: mach_header.self)
            addBinaryImageFunc?(header, 0)
            XCTAssertEqual(6, imageCache.cache?.count ?? 0)
        }
        
        if machHeadersExpectArray.count > 1 {
            let header1 = machHeadersExpectArray[1].assumingMemoryBound(to: mach_header.self)
            removeBinaryImageFunc?(header1, 0)
        }
        
        if machHeadersExpectArray.count > 2 {
            let header2 = machHeadersExpectArray[2].assumingMemoryBound(to: mach_header.self)
            removeBinaryImageFunc?(header2, 0)
        }
        
        XCTAssertEqual(4, imageCache.cache?.count ?? 0)
        imageCache.stop()
        
        if machHeadersTestCache.count > 6 {
            let header = machHeadersTestCache[6].assumingMemoryBound(to: mach_header.self)
            addBinaryImageFunc?(header, 0)
        }
        XCTAssertNil(imageCache.cache)
    }
    
    // MARK: - Helper Methods
    
    private func assertBinaryImageCacheLength(_ expected: Int) {
        var counter = 0
        withUnsafeMutablePointer(to: &counter) { counterPtr in
            sentrycrashbic_iterateOverImages(countNumberOfImagesInCache, UnsafeMutableRawPointer(counterPtr))
        }
        XCTAssertEqual(counter, expected)
    }
    
    private func assertCachedBinaryImages() {
        let cached = binaryImageCacheToArray()
        for i in 0..<cached.count {
            guard let binaryImageValue = cached[i] as? NSValue else { continue }
            let binaryImage = binaryImageValue.pointerValue!.assumingMemoryBound(to: SentryCrashBinaryImage.self)
            let header = machHeadersExpectArray[i]
            XCTAssertEqual(binaryImage.pointee.address, UInt64(bitPattern: Int64(Int(bitPattern: header))))
        }
    }
    
    private func binaryImageCacheToArray() -> NSArray {
        let result = NSMutableArray()
        let unmanagedArray = Unmanaged.passUnretained(result)
        sentrycrashbic_iterateOverImages(addBinaryImageToArray, unmanagedArray.toOpaque())
        return result
    }
}
